error-level metadb noise for cosign v3 sigstore-bundle referrers (parse.go getCosignSignatureLayersInfo); DELETE of such referrers returns 500 though deletion succeeds

### zot version

v2.1.18 (Kubernetes, StatefulSet, local storage, retention + basic-auth ACL config)

### Describe the bug

cosign v3.x signs images as OCI 1.1 referrers carrying `application/vnd.dev.sigstore.bundle.v0.3+json` artifacts (no `sha256-<digest>.sig` tag, no `dev.cosignproject.cosign/signature` layer annotation). Two rough edges in zot's metadb when these are present:

**1. Error-level log noise per signed image (cosmetic).** Every metadb parse of a repo holding such a referrer emits ~2 error-level lines per signature:

```json
{"level":"error","message":"failed to get specific annotation of cosign signature","repository":"parascope-vault-collector","reference":"sha256:fea71caf17b2776a305a16fff61b54fb61303ecf9e5036f01362ba88c5dc400b","layerDigest":"sha256:e7c62a51d4e1993b50c4a33089c76f757069df01fe49f094015679977aabf38c","caller":"zotregistry.dev/zot/v2/pkg/meta/parse.go:395","func":"zotregistry.dev/zot/v2/pkg/meta.getCosignSignatureLayersInfo"}
```

`getCosignSignatureLayersInfo` appears to look for the legacy cosign layer annotation, which sigstore-bundle referrer layers legitimately do not carry. Everything works — the referrers API serves them, `cosign verify` succeeds against the referrer — so this looks purely cosmetic, but at error level it pollutes alerting pipelines that key on log severity. A signature-format-aware downgrade to debug/info (or handling the bundle format's annotations) would fix it.

**2. `DELETE /v2/<repo>/manifests/<referrer-digest>` returns HTTP 500 — but the deletion succeeds (functional wart).** Deleting a sigstore-bundle referrer manifest by digest (credential has delete permission) returns 500, yet a subsequent `GET /v2/<repo>/referrers/<subject-digest>` shows the referrer gone and the manifest is no longer retrievable. Correlated log line at the moment of each 500:

```json
{"level":"error","message":"failed to check if image is a signature or not","error":"image meta not found","component":"metadb","caller":"zotregistry.dev/zot/v2/pkg/meta/hooks.go:243","func":"zotregistry.dev/zot/v2/pkg/meta.OnDeleteManifest"}
```

So the storage delete goes through and the `OnDeleteManifest` metadb hook then fails to classify the just-deleted manifest ("image meta not found"), which surfaces as a 500 to the client despite the operation having taken effect. Clients doing cleanup have to treat 500 as "re-query to find out".

### To reproduce

1. Push any image: `crane append --oci-empty-base --new_layer layer.tar.gz --new_tag <zot>/repo:tag`
2. Sign with cosign v3 (v3.1.2 here): `cosign sign --yes --key k.key --tlog-upload=false --use-signing-config=false <zot>/repo@<digest>` → referrer lands, works.
3. Watch the zot log: the parse.go:395 error lines appear when metadb parses the repo (startup parse, and on access).
4. `curl -X DELETE <zot>/v2/repo/manifests/<referrer-manifest-digest>` with a delete-capable user → HTTP 500, hooks.go:243 error logged, but the referrer is actually gone on re-query.

### Expected behavior

1. No error-level logging for well-formed sigstore-bundle referrers (they are not malformed cosign signatures — they are the new format).
2. DELETE of a referrer manifest that succeeds returns 202, not 500.

### Additional context

Both behaviors observed on a production registry; happy to provide fuller logs or re-run repro steps against a scratch repo if useful.

--- a/errors/errors.go
+++ b/errors/errors.go
@@ -109,6 +109,7 @@ var (
 	ErrBadRange                         = errors.New("bad range for streaming blob")
 	ErrBadLayerCount                    = errors.New("manifest layers count doesn't correspond to config history")
 	ErrManifestConflict                 = errors.New("multiple manifests found")
+	ErrManifestRestoreAttempted         = errors.New("manifest restore to image store attempted after a metadb failure")
 	ErrImageMetaNotFound                = errors.New("image meta not found")
 	ErrUnexpectedMediaType              = errors.New("unexpected media type")
 	ErrRepoMetaNotFound                 = errors.New("repo metadata not found for given repo name")

--- a/pkg/api/routes.go
+++ b/pkg/api/routes.go
@@ -1007,9 +1007,15 @@ func (rh *RouteHandler) DeleteManifest(response http.ResponseWriter, request *ht
 		err := meta.OnDeleteManifest(name, reference, mediaType, manifestDigest, manifestBlob,
 			rh.c.StoreController, rh.c.MetaDB, rh.c.Log)
 		if err != nil {
-			response.WriteHeader(http.StatusInternalServerError)
+			// The image store manifest may have been restored (or restore may itself have
+			// failed), so this must not report success either way.
+			if errors.Is(err, zerr.ErrManifestRestoreAttempted) {
+				rh.c.Log.Error().Err(err).Str("repository", name).Str("reference", reference).
+					Msg("failed to remove manifest from metadb; attempted to restore it in the image store")
+				response.WriteHeader(http.StatusInternalServerError)
 
-			return
+				return
+			}
 		}
 	}
 

--- a/pkg/api/routes_test.go
+++ b/pkg/api/routes_test.go
@@ -34,9 +34,13 @@ import (
 	"zotregistry.dev/zot/v2/pkg/api/config"
 	"zotregistry.dev/zot/v2/pkg/api/constants"
 	apiErr "zotregistry.dev/zot/v2/pkg/api/errors"
+	zcommon "zotregistry.dev/zot/v2/pkg/common"
+	"zotregistry.dev/zot/v2/pkg/extensions/monitoring"
 	"zotregistry.dev/zot/v2/pkg/log"
+	"zotregistry.dev/zot/v2/pkg/meta/boltdb"
 	mTypes "zotregistry.dev/zot/v2/pkg/meta/types"
 	reqCtx "zotregistry.dev/zot/v2/pkg/requestcontext"
+	"zotregistry.dev/zot/v2/pkg/storage/local"
 	storageTypes "zotregistry.dev/zot/v2/pkg/storage/types"
 	test "zotregistry.dev/zot/v2/pkg/test/common"
 	"zotregistry.dev/zot/v2/pkg/test/mocks"
@@ -3538,4 +3542,193 @@ func TestGetBlobMultipartReaderCloseError(t *testing.T) {
 	body := drainResponseBody(t, resp)
 	assert.Less(t, int64(len(body)), advertisedLen,
 		"a Close() error on the second range must truncate the body")
+}
+
+// TestDeleteManifestSucceedsDespiteMetaDBHookFailure verifies that an OnDeleteManifest failure
+// after a successful image-store delete does not surface as a client-facing error.
+func TestDeleteManifestSucceedsDespiteMetaDBHookFailure(t *testing.T) {
+	Convey("Deleting a signature referrer whose metaDB cleanup fails still returns 202", t, func() {
+		subjectDigest := godigest.FromString("subject-manifest")
+
+		manifest := ispec.Manifest{
+			Versioned:    specs.Versioned{SchemaVersion: 2},
+			MediaType:    ispec.MediaTypeImageManifest,
+			ArtifactType: zcommon.ArtifactTypeCosignBundle,
+			Config: ispec.Descriptor{
+				MediaType: "application/vnd.oci.empty.v1+json",
+				Digest:    godigest.FromString("empty-config"),
+				Size:      2,
+			},
+			Subject: &ispec.Descriptor{
+				MediaType: ispec.MediaTypeImageManifest,
+				Digest:    subjectDigest,
+				Size:      10,
+			},
+		}
+
+		manifestJSON, err := json.Marshal(manifest)
+		So(err, ShouldBeNil)
+
+		referrerDigest := godigest.FromBytes(manifestJSON)
+
+		store := mocks.MockedImageStore{
+			GetImageManifestFn: func(repo, reference string) ([]byte, godigest.Digest, string, error) {
+				return manifestJSON, referrerDigest, ispec.MediaTypeImageManifest, nil
+			},
+			DeleteImageManifestFn: func(ctx context.Context, repo, reference string, detectCollision bool) error {
+				return nil
+			},
+		}
+
+		ctlr := api.NewController(config.New())
+		ctlr.Router = mux.NewRouter()
+		ctlr.StoreController.DefaultStore = store
+		ctlr.MetaDB = mocks.MetaDBMock{
+			DeleteSignatureFn: func(repo string, signedManifestDigest godigest.Digest,
+				sigMeta mTypes.SignatureMetadata,
+			) error {
+				// Simulates a concurrent writer already removing this entry.
+				return zerr.ErrImageMetaNotFound
+			},
+		}
+
+		handler := api.NewRouteHandler(ctlr)
+
+		req := httptest.NewRequestWithContext(
+			context.Background(),
+			http.MethodDelete,
+			"http://example.com/v2/test/manifests/"+referrerDigest.String(),
+			http.NoBody,
+		)
+		req = mux.SetURLVars(req, map[string]string{
+			"name":      "test",
+			"reference": referrerDigest.String(),
+		})
+
+		rec := httptest.NewRecorder()
+		handler.DeleteManifest(rec, req)
+
+		resp := rec.Result()
+		defer resp.Body.Close()
+
+		So(resp.StatusCode, ShouldEqual, http.StatusAccepted)
+	})
+}
+
+// TestDeleteManifestSucceedsAfterSignatureSubjectOrphaned runs the same scenario end to end
+// against real components (no mocked metaDB).
+func TestDeleteManifestSucceedsAfterSignatureSubjectOrphaned(t *testing.T) {
+	Convey("Deleting a signature referrer whose subject was already untagged", t, func() {
+		rootDir := t.TempDir()
+		testLog := log.NewTestLogger()
+		metrics := monitoring.NewMetricsServer(false, testLog)
+		defer metrics.Stop()
+
+		imgStore := local.NewImageStore(rootDir, true, true, testLog, metrics, nil, nil, nil, nil)
+
+		boltDriver, err := boltdb.GetBoltDriver(boltdb.DBParameters{RootDir: rootDir})
+		So(err, ShouldBeNil)
+
+		metaDB, err := boltdb.New(boltDriver, testLog)
+		So(err, ShouldBeNil)
+
+		ctx := context.Background()
+
+		ctlr := api.NewController(config.New())
+		ctlr.Router = mux.NewRouter()
+		ctlr.StoreController.DefaultStore = imgStore
+		ctlr.MetaDB = metaDB
+		handler := api.NewRouteHandler(ctlr)
+
+		putReq := func(reference string, body []byte) int {
+			req := httptest.NewRequestWithContext(ctx, http.MethodPut,
+				"http://example.com/v2/repo/manifests/"+reference, bytes.NewReader(body))
+			req.Header.Set("Content-Type", ispec.MediaTypeImageManifest)
+			req = mux.SetURLVars(req, map[string]string{"name": "repo", "reference": reference})
+
+			rec := httptest.NewRecorder()
+			handler.UpdateManifest(rec, req)
+			rec.Result().Body.Close()
+
+			return rec.Result().StatusCode
+		}
+
+		deleteReq := func(reference string) int {
+			req := httptest.NewRequestWithContext(ctx, http.MethodDelete,
+				"http://example.com/v2/repo/manifests/"+reference, http.NoBody)
+			req = mux.SetURLVars(req, map[string]string{"name": "repo", "reference": reference})
+
+			rec := httptest.NewRecorder()
+			handler.DeleteManifest(rec, req)
+			rec.Result().Body.Close()
+
+			return rec.Result().StatusCode
+		}
+
+		// Push and tag the subject image through the real hook path.
+		subjectManifest := ispec.Manifest{
+			Versioned: specs.Versioned{SchemaVersion: 2},
+			MediaType: ispec.MediaTypeImageManifest,
+			Config: ispec.Descriptor{
+				MediaType: ispec.MediaTypeImageConfig,
+				Digest:    godigest.FromString("{}"),
+				Size:      2,
+			},
+			Layers: []ispec.Descriptor{},
+		}
+		subjectBody, err := json.Marshal(subjectManifest)
+		So(err, ShouldBeNil)
+		subjectDigest := godigest.FromBytes(subjectBody)
+
+		_, _, err = imgStore.FullBlobUpload(ctx, "repo", bytes.NewReader([]byte("{}")), godigest.FromString("{}"))
+		So(err, ShouldBeNil)
+
+		statusCode := putReq("latest", subjectBody)
+		So(statusCode, ShouldEqual, http.StatusCreated)
+
+		// Sign it: push a sigstore-bundle referrer whose Subject is the subject image.
+		emptyConfig := []byte("{}")
+		emptyConfigDigest := godigest.FromBytes(emptyConfig)
+		_, _, err = imgStore.FullBlobUpload(ctx, "repo", bytes.NewReader(emptyConfig), emptyConfigDigest)
+		So(err, ShouldBeNil)
+
+		referrer := ispec.Manifest{
+			Versioned:    specs.Versioned{SchemaVersion: 2},
+			MediaType:    ispec.MediaTypeImageManifest,
+			ArtifactType: zcommon.ArtifactTypeCosignBundle,
+			Config: ispec.Descriptor{
+				MediaType: "application/vnd.oci.empty.v1+json",
+				Digest:    emptyConfigDigest,
+				Size:      int64(len(emptyConfig)),
+			},
+			Layers: []ispec.Descriptor{},
+			Subject: &ispec.Descriptor{
+				MediaType: ispec.MediaTypeImageManifest,
+				Digest:    subjectDigest,
+				Size:      int64(len(subjectBody)),
+			},
+		}
+		referrerBody, err := json.Marshal(referrer)
+		So(err, ShouldBeNil)
+		referrerDigest := godigest.FromBytes(referrerBody)
+
+		statusCode = putReq(referrerDigest.String(), referrerBody)
+		So(statusCode, ShouldEqual, http.StatusCreated)
+
+		repoMeta, err := metaDB.GetRepoMeta(ctx, "repo")
+		So(err, ShouldBeNil)
+		So(repoMeta.Signatures, ShouldContainKey, subjectDigest.String())
+
+		// Delete the subject's only tag: its digest is now untagged.
+		statusCode = deleteReq("latest")
+		So(statusCode, ShouldEqual, http.StatusAccepted)
+
+		repoMeta, err = metaDB.GetRepoMeta(ctx, "repo")
+		So(err, ShouldBeNil)
+		So(repoMeta.Signatures, ShouldNotContainKey, subjectDigest.String())
+
+		// Deleting the now-orphaned referrer must still return 202.
+		statusCode = deleteReq(referrerDigest.String())
+		So(statusCode, ShouldEqual, http.StatusAccepted)
+	})
 }

--- a/pkg/extensions/search/search_test.go
+++ b/pkg/extensions/search/search_test.go
@@ -6583,7 +6583,9 @@ func TestMetaDBWhenDeletingImages(t *testing.T) {
 				resp, err = resty.R().Delete(baseURL + "/v2/" + "repo1" + "/manifests/" + "signatureReference")
 				So(resp, ShouldNotBeNil)
 				So(err, ShouldBeNil)
-				So(resp.StatusCode(), ShouldEqual, http.StatusInternalServerError)
+				// the image is already gone from the image store at this point, so a metadb
+				// bookkeeping failure here is not exposed to the client.
+				So(resp.StatusCode(), ShouldEqual, http.StatusAccepted)
 			})
 
 			Convey("image is a signature, DeleteSignature fails", func() {
@@ -6614,10 +6616,12 @@ func TestMetaDBWhenDeletingImages(t *testing.T) {
 					"sha256-343ebab94a7674da181c6ea3da013aee4f8cbe357870f8dcaf6268d5343c3474.sig")
 				So(resp, ShouldNotBeNil)
 				So(err, ShouldBeNil)
-				So(resp.StatusCode(), ShouldEqual, http.StatusInternalServerError)
+				// the image is already gone from the image store at this point, so a metadb
+				// bookkeeping failure here is not exposed to the client.
+				So(resp.StatusCode(), ShouldEqual, http.StatusAccepted)
 			})
 
-			Convey("image is a signature, PutImageManifest fails", func() {
+			Convey("non-signature delete, RemoveRepoReference and restore PutImageManifest both fail", func() {
 				ctlr.StoreController.DefaultStore = mocks.MockedImageStore{
 					NewBlobUploadFn: ctlr.StoreController.DefaultStore.NewBlobUpload,
 					PutBlobChunkFn:  ctlr.StoreController.DefaultStore.PutBlobChunk,
@@ -6652,6 +6656,7 @@ func TestMetaDBWhenDeletingImages(t *testing.T) {
 					"343ebab94a7674da181c6ea3da013aee4f8cbe357870f8dcaf6268d5343c3474.sig")
 				So(resp, ShouldNotBeNil)
 				So(err, ShouldBeNil)
+				// A restore may have rolled back the delete, so this must be a 500.
 				So(resp.StatusCode(), ShouldEqual, http.StatusInternalServerError)
 			})
 		})

--- a/pkg/meta/hooks.go
+++ b/pkg/meta/hooks.go
@@ -3,6 +3,7 @@ package meta
 import (
 	"context"
 	"errors"
+	"fmt"
 	"slices"
 
 	godigest "github.com/opencontainers/go-digest"
@@ -240,8 +241,15 @@ func OnDeleteManifest(repo, reference, mediaType string, digest godigest.Digest,
 			SignatureType:   signatureType,
 		})
 		if err != nil {
-			log.Error().Err(err).Str("component", "metadb").
-				Msg("failed to check if image is a signature or not")
+			if errors.Is(err, zerr.ErrImageMetaNotFound) {
+				// Expected: RemoveRepoReference already deleted Signatures[signedManifestDigest]
+				// when the last reference to the signed manifest was removed.
+				log.Debug().Err(err).Str("component", "metadb").
+					Msg("signature meta already removed")
+			} else {
+				log.Error().Err(err).Str("component", "metadb").
+					Msg("failed to delete signature meta")
+			}
 
 			manageRepoMetaSuccessfully = false
 		}
@@ -264,6 +272,10 @@ func OnDeleteManifest(repo, reference, mediaType string, digest godigest.Digest,
 	if !manageRepoMetaSuccessfully {
 		log.Info().Str("tag", reference).Str("repository", repo).Str("component", "metadb").
 			Msg("failed to delete image meta was unsuccessful for tag in repo")
+
+		if !isSignature {
+			return fmt.Errorf("%w: %w", zerr.ErrManifestRestoreAttempted, err)
+		}
 
 		return err
 	}

--- a/pkg/meta/hooks_test.go
+++ b/pkg/meta/hooks_test.go
@@ -1,7 +1,9 @@
 package meta_test
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"errors"
 	"testing"
 
@@ -10,6 +12,7 @@ import (
 	. "github.com/smartystreets/goconvey/convey"
 
 	zerr "zotregistry.dev/zot/v2/errors"
+	zcommon "zotregistry.dev/zot/v2/pkg/common"
 	"zotregistry.dev/zot/v2/pkg/extensions/monitoring"
 	"zotregistry.dev/zot/v2/pkg/log"
 	"zotregistry.dev/zot/v2/pkg/meta"
@@ -467,4 +470,91 @@ func TestUpdateErrors(t *testing.T) {
 			So(err, ShouldBeNil)
 		})
 	})
+}
+
+// TestOnDeleteManifest_OrphanedSignatureReferrerDeletionSucceeds verifies that
+// OnDeleteManifest returns ErrImageMetaNotFound when deleting a signature referrer
+// whose subject manifest was already untagged.
+func TestOnDeleteManifest_OrphanedSignatureReferrerDeletionSucceeds(t *testing.T) {
+	Convey("Deleting a signature referrer after its subject's last tag is gone fails with ErrImageMetaNotFound",
+		t, func() {
+			rootDir := t.TempDir()
+			storeController := storage.StoreController{}
+			log := log.NewTestLogger()
+			metrics := monitoring.NewMetricsServer(false, log)
+
+			defer metrics.Stop()
+			storeController.DefaultStore = local.NewImageStore(rootDir, true, true, log, metrics, nil, nil, nil, nil)
+
+			params := boltdb.DBParameters{RootDir: rootDir}
+			boltDriver, err := boltdb.GetBoltDriver(params)
+			So(err, ShouldBeNil)
+
+			metaDB, err := boltdb.New(boltDriver, log)
+			So(err, ShouldBeNil)
+
+			imgStore := storeController.GetImageStore("repo")
+			ctx := context.Background()
+
+			// Push and register the subject image.
+			image := CreateDefaultImage()
+			subjectDigest := image.Digest()
+			subjectBody := image.ManifestDescriptor.Data
+
+			So(WriteImageToFileSystem(image, "repo", "latest", storeController), ShouldBeNil)
+			So(meta.OnUpdateManifest(ctx, "repo", "latest", ispec.MediaTypeImageManifest, subjectDigest, subjectBody,
+				storeController, metaDB, log), ShouldBeNil)
+
+			// Push and register a sigstore-bundle referrer signing it.
+			emptyConfig := []byte("{}")
+			emptyConfigDigest := godigest.FromBytes(emptyConfig)
+			_, _, err = imgStore.FullBlobUpload(ctx, "repo", bytes.NewReader(emptyConfig), emptyConfigDigest)
+			So(err, ShouldBeNil)
+
+			referrer := ispec.Manifest{
+				MediaType:    ispec.MediaTypeImageManifest,
+				ArtifactType: zcommon.ArtifactTypeCosignBundle,
+				Config: ispec.Descriptor{
+					MediaType: "application/vnd.oci.empty.v1+json",
+					Digest:    emptyConfigDigest,
+					Size:      int64(len(emptyConfig)),
+				},
+				Layers: []ispec.Descriptor{},
+				Subject: &ispec.Descriptor{
+					MediaType: ispec.MediaTypeImageManifest,
+					Digest:    subjectDigest,
+					Size:      int64(len(subjectBody)),
+				},
+			}
+			referrer.SchemaVersion = 2
+
+			referrerBody, err := json.Marshal(referrer)
+			So(err, ShouldBeNil)
+			referrerDigest := godigest.FromBytes(referrerBody)
+
+			_, _, err = imgStore.PutImageManifest(ctx, "repo", referrerDigest.String(), ispec.MediaTypeImageManifest,
+				referrerBody, nil)
+			So(err, ShouldBeNil)
+			So(meta.OnUpdateManifest(ctx, "repo", referrerDigest.String(), ispec.MediaTypeImageManifest, referrerDigest,
+				referrerBody, storeController, metaDB, log), ShouldBeNil)
+
+			repoMeta, err := metaDB.GetRepoMeta(ctx, "repo")
+			So(err, ShouldBeNil)
+			So(repoMeta.Signatures, ShouldContainKey, subjectDigest.String())
+
+			// Delete the subject's only tag.
+			So(imgStore.DeleteImageManifest(ctx, "repo", "latest", false), ShouldBeNil)
+			So(meta.OnDeleteManifest("repo", "latest", ispec.MediaTypeImageManifest, subjectDigest, subjectBody,
+				storeController, metaDB, log), ShouldBeNil)
+
+			repoMeta, err = metaDB.GetRepoMeta(ctx, "repo")
+			So(err, ShouldBeNil)
+			So(repoMeta.Signatures, ShouldNotContainKey, subjectDigest.String())
+
+			// The referrer is still present in the image store; deleting it now fails.
+			So(imgStore.DeleteImageManifest(ctx, "repo", referrerDigest.String(), false), ShouldBeNil)
+			err = meta.OnDeleteManifest("repo", referrerDigest.String(), ispec.MediaTypeImageManifest, referrerDigest,
+				referrerBody, storeController, metaDB, log)
+			So(errors.Is(err, zerr.ErrImageMetaNotFound), ShouldBeTrue)
+		})
 }

--- a/pkg/meta/parse.go
+++ b/pkg/meta/parse.go
@@ -399,10 +399,13 @@ func getCosignSignatureLayersInfo(
 			return layers, err
 		}
 
+		// The legacy cosign annotation is absent on sigstore-bundle referrers (and any other
+		// non-legacy cosign format) by design. SignatureKey is simply left
+		// empty below in that case.
 		layerSigKey, ok := layer.Annotations[zcommon.CosignSigKey]
 		if !ok {
-			log.Error().Err(err).Str("repository", repo).Str("reference", tag).Str("layerDigest", layer.Digest.String()).Msg(
-				"failed to get specific annotation of cosign signature")
+			log.Debug().Str("repository", repo).Str("reference", tag).Str("layerDigest", layer.Digest.String()).Msg(
+				"cosign signature layer has no legacy signature-key annotation")
 		}
 
 		layers = append(layers, mTypes.LayerInfo{

--- a/test/blackbox/annotations.bats
+++ b/test/blackbox/annotations.bats
@@ -189,6 +189,58 @@ function teardown_file() {
     [[ "$sigName" == *"${digest}"* ]]
 }
 
+@test "delete cosign sigstore-bundle referrer returns 202 without error-level logs" {
+    zot_port=`cat ${BATS_FILE_TMPDIR}/zot.port`
+
+    # Dedicated image content so this digest starts with zero referrers.
+    cat > ${BATS_FILE_TMPDIR}/Dockerfile.delete-referrer-test<<EOF
+FROM public.ecr.aws/t0x7q1g8/centos:7
+CMD ["/bin/sh", "-c", "echo 'delete cosign referrer test'"]
+EOF
+    run podman build -f ${BATS_FILE_TMPDIR}/Dockerfile.delete-referrer-test -t 127.0.0.1:${zot_port}/annotations:delete-referrer-test . --format oci
+    [ "$status" -eq 0 ]
+    run podman push 127.0.0.1:${zot_port}/annotations:delete-referrer-test --tls-verify=false --format=oci
+    [ "$status" -eq 0 ]
+
+    run curl -s -I -H "Accept: application/vnd.oci.image.manifest.v1+json" http://localhost:${zot_port}/v2/annotations/manifests/delete-referrer-test
+    [ "$status" -eq 0 ]
+    local digest=$(echo "${output}" | grep -i docker-content-digest | tr -d '\r' | awk '{print $2}')
+    [ -n "$digest" ]
+
+    run cosign initialize
+    [ "$status" -eq 0 ]
+    run cosign generate-key-pair --output-key-prefix "${BATS_FILE_TMPDIR}/cosign-sign-test-delete"
+    [ "$status" -eq 0 ]
+    run env COSIGN_EXPERIMENTAL=1 cosign sign --new-bundle-format=true --registry-referrers-mode=oci-1-1 --key ${BATS_FILE_TMPDIR}/cosign-sign-test-delete.key --allow-insecure-registry localhost:${zot_port}/annotations@${digest} --yes
+    [ "$status" -eq 0 ]
+
+    run curl -s http://localhost:${zot_port}/v2/annotations/referrers/${digest}
+    [ "$status" -eq 0 ]
+    local referrerDigest=$(echo "${output}" | jq -r '.manifests[0].digest')
+    [ -n "$referrerDigest" ]
+    [ "$referrerDigest" != "null" ]
+
+    # A well-formed sigstore-bundle referrer must never log the missing legacy
+    # annotation at error level.
+    run bash -c "grep -c '\"level\":\"error\".*legacy signature-key annotation' ${BATS_FILE_TMPDIR}/zot.log || true"
+    [ "$status" -eq 0 ]
+    [ "${output}" -eq 0 ]
+
+    # Delete the subject's only tag; the referrer above is left untouched, undeleted.
+    run curl -s -o /dev/null -w "%{http_code}" -X DELETE "http://localhost:${zot_port}/v2/annotations/manifests/delete-referrer-test"
+    [ "$status" -eq 0 ]
+    [ "${lines[-1]}" -eq 202 ]
+
+    # Deleting the now-orphaned referrer must still return 202, not 500.
+    run curl -s -o /dev/null -w "%{http_code}" -X DELETE "http://localhost:${zot_port}/v2/annotations/manifests/${referrerDigest}"
+    [ "$status" -eq 0 ]
+    [ "${lines[-1]}" -eq 202 ]
+
+    run curl -s http://localhost:${zot_port}/v2/annotations/referrers/${digest}
+    [ "$status" -eq 0 ]
+    [ $(echo "${output}" | jq --arg d "${referrerDigest}" '[.manifests[] | select(.digest == $d)] | length') -eq 0 ]
+}
+
 @test "zot reports IsSigned true for cosign referrer signatures" {
     zot_port=`cat ${BATS_FILE_TMPDIR}/zot.port`
     run curl -X POST -H "Content-Type: application/json" --data '{ "query": "{ ImageList(repo: \"annotations\") { Results { RepoName Tag Manifests {Digest ConfigDigest Size Layers { Size Digest }} Vendor Licenses }}}"}' http://localhost:${zot_port}/v2/_zot/ext/search

